BUG: Fix issue with incorrect groupby handling of NaT

larvian opened the pull request:
closes #10590 

For groupby the time stamps gets converted to integervalue `tslib.iNaT`
which is -9223372036854775808. The aggregation is then done using this
value with incorrect result as a consequence. The solution proposed here
is to replace its value by np.nan in case it is a `datetime64[ns]`

jreback commented:
test would be nice

larvian commented:
I have tried to add a test and update to doc. I am not very sure I have done this correctly :)

larvian commented:
I don't know what the next step here is? Am I supposed to do something now or am I waiting for someone to have a look at what I have done? There is a **This branch has conflicts that must be resolved** Can I see what the conflicts are and fix them?

maximilianr commented:
@larvian rebase your branch on upstream master to resolve the conflicts

larvian commented:
@maximilianr Thanks! I will try to figure out how to do that :)

jreback commented:
lots of tips [here](http://pandas.pydata.org/pandas-docs/stable/contributing.html)

sinhrks commented:
Also, can you add tests for ``timedelta`` which looks affected also?

```
df = pd.DataFrame({'IX': ['A', 'A'], 'num': [np.nan, 100],
                  'dt': [np.nan, datetime.datetime.now()],
                  'td':[np.nan, datetime.timedelta(days=1)]})
# OK
df.min()
# IX                              A
# dt    2015-07-25 07:41:47.871654
# num                           100
# td                1 days 00:00:00
# dtype: object

# NG
# df.groupby('IX').min()
#     dt  num  td
# IX             
# A  NaT  100 NaT
```

I *think* you can use ``com.is_datetime_or_timedelta_dtype`` to check dtype.

- https://github.com/pydata/pandas/blob/master/pandas/core/common.py#L2517

larvian commented:
@jreback Thanks!
@sinhrks Ok I will look at that too.

larvian commented:
Hm... I can't get this rebase to work. It gives conflicts which seems to come back after I fix :(
It seems to get messier the more I try.
Perhaps it is easiest to create a new pandas clone and apply my changes there?
I have to leave it for another day.

TomAugspurger commented:
We can help walk you through the git problems. Could be easier in the chat room: https://gitter.im/pydata/pandas

larvian commented:
Thanks Tom!
I think I need it :)

larvian committed:
BUG: Fix issue with incorrect groupby handling of NaT #10590

For groupby the time stamps gets converted to integervalue tslib.iNaT
which is -9223372036854775808. The aggregation is then done using this
value with incorrect result as a consequence. The solution proposed here
is to replace its value by np.nan in case it is a datetime or timedelta.

larvian commented:
Just curious, what's next with this PR?
For now is there anything else I should do?

jreback commented on pandas/core/groupby.py:
```diff
@@ -2608,6 +2608,8 @@ def _cython_agg_blocks(self, how, numeric_only=True):
         for block in data.blocks:
 
             values = block._try_operate(block.values)
```
I'd like to do this differently.

move the ``._try_operate`` call to inse ``.aggregate`` (its line 1533). Then you can use do the ``iNaT`` to ``np.nan`` there

larvian replied:
Ok,
Then I assume I would do the `._try_operate` after checking `is_datetime_or_timedelta_dtype` not to lose dtype.
But `._try_operate` on `DatetimeBlock` is just `values.view('i8')`
However the code just after `is_datetime_or_timedelta_dtype` is `values = values.view('int64')`
I assume both are not needed so I can remove `._try_operate`
Hope this doesn't destroy something else.

jreback commented on pandas/tests/test_groupby.py:
```diff
@@ -5313,6 +5313,20 @@ def test_func(x):
+    def test_first_last_max_min_on_time_data(self):
+        from datetime import timedelta as td
+        DF_dt_test=DataFrame({'dt':[nan,'2015-07-24 10:10','2015-07-25 11:11','2015-07-23 12:12',nan],'td':[nan,td(days=1) ,td(days=2),td(days=3),nan]})
+        DF_dt_test.dt=pd.to_datetime(DF_dt_test.dt)
```
wrap this line
use df as a variable name

jreback replied:
add the issue number as a comment

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -613,3 +613,4 @@ Bug Fixes
 
 - Bug in ``pd.DataFrame`` when constructing an empty DataFrame with a string dtype (:issue:`9428`)
 - Bug in ``pd.unique`` for arrays with the ``datetime64`` or ``timedelta64`` dtype that meant an array with object dtype was returned instead the original dtype (:issue: `9431`)
```
dont reference internal functions
just say that bug in groupy with NaT were computing incorrectly 
someone who wants more detail can always drill down

jreback commented:
can you rebase / update

larvian commented:
Sure... rebase done. Queued for Travis CI

larvian commented:
Rebased yesterday again and it passed Travis CI.
Today again I have "This branch has conflicts that must be resolved".
This seem to happen all the time due to changes to v0.17.0.txt .
It seems wasteful of Travis time to run CI just because updates due to changes to v0.17.0.txt.
How to avoid the constant conflicts due to changes in this file?
This time I try to insert my line above the last line.

jreback commented:
to avoid conflicts, don't insert the line at the end of the notes, put it somewhere else above. This is why I have left lots of blank space (which gradually gets filled up).

jreback commented:
ping when green.

larvian commented:
This one is green now

jreback commented:
thanks!

behzadnouri commented:
[this added line](https://github.com/pydata/pandas/pull/10625/files#diff-720d374f1a709d0075a1f0a02445cd65R1535) has broken cython aggregation. You cannot assign `nan` to an `int64` view of an array.

before this patch:

```python
In [1]: from string import ascii_lowercase

In [2]: np.random.seed(2718281)

In [3]: n = 1 << 21

In [4]: dr = date_range('2015-08-30', periods=n // 10, freq='T')

In [5]: df = DataFrame({
   ...:         '1st':np.random.choice(list(ascii_lowercase), n),
   ...:         '2nd':np.random.randint(0, 5, n),
   ...:         '3rd':np.random.choice(dr, n)})

In [6]: df.loc[np.random.choice(n, n // 10), '3rd'] = np.nan

In [7]: gr = df.groupby(['1st', '2nd'])

In [8]: %timeit gr.count()
The slowest run took 21.22 times longer than the fastest. This could mean that an intermediate result is being cached 
1 loops, best of 3: 13.3 ms per loop

In [9]: %timeit gr.count()
100 loops, best of 3: 13.8 ms per loop

In [10]: pd.__version__
Out[10]: '0.16.2+521.g207efc2'
```

with this patch:

```python
In [8]: %timeit gr.count()
1 loops, best of 3: 144 ms per loop

In [9]: %timeit gr.count()
10 loops, best of 3: 149 ms per loop

In [10]: pd.__version__
Out[10]: '0.16.2+522.g9c2d1a6'
```

jreback commented:
yes that should be iNaT

odd that the tests didn't pick this up

behzadnouri commented:
@jreback it is modifying values which are already `iNaT`:

    values[values == tslib.iNaT] = np.nan

jreback commented:
this should already be a float here

jreback commented:
ahh I see it is raising, being cause and taking the slow path...ok.thanks for the catch

larvian commented:
@behzadnouri @jreback Thanks for pointing out the problem.
Is it that the cython aggregation function accepts `int64` does not like the `nan` and raises exception?
Not sure how to deal with it. Should all the `NaT` be dropped instead before aggregating?
Are there some aggregations where the `NaT` are important?
